lib/spack/spack/solver/asp.py

```diff
@@ -1566,9 +1566,20 @@ class Body(object):
                         if concrete_build_deps or dtype != "build":
                             clauses.append(fn.attr("depends_on", spec.name, dep.name, dtype))
 
-                            # Ensure Spack will not coconcretize this with another provider
-                            # for the same virtual
-                            for virtual in dep.package.virtuals_provided:
+                            # Skip virtual node constriants for renamed/deleted packages,
+                            # so their binaries can still be installed.
+                            # NOTE: with current specs (which lack edge attributes) this
+                            # can allow concretizations with two providers, but it's unlikely.
```

Comment:
This is a bit scary :worried: Not sure what we could do more than this though. I agree it's unlikely, since one of the provider must be in a buildcache and not in the current repository.

Reply:
I think we don't rename things (especially MPI and other virtuals) frequently enough for this to matter.  Though @eugeneswalker and the E4S team are trying hard to make it matter with their 100,000 package buildcache.

Reply:
I think the value of reuse here outweighs the corner case.